### PR TITLE
File upload in Java.

### PR DESCRIPTION
**Pull request by 23Skidoo**
Things left to implement:

- [x] Download index generation.
- [x] File downloading.
- [x] Support the `path` parameter.
- [x] Check that paths that include `..` are correctly handled.
- [x] Test integration with desktop client.

### CODE REVIEW

**Comment by 23Skidoo** [comment]
To test, access `localhost:7000/upload` via browser.

**Comment by 23Skidoo** [comment]
Implemented file download.

**Comment by 23Skidoo** [comment]
Implemented directory index generation.

File: src/server/Doorman.java [hunk 1]
```diff
@@ -56,6 +60,9 @@
 	public Doorman(CommandHandler commandHandler, int port) throws IOException {
 
 		this.commandHandler = commandHandler;
+		// TODO: Don't hard-code this location.
```
Comment (tastyburger) [inline src/server/Doorman.java]: There is actually a constant identifier for the resource path in Executor but it's not widely used yet.

File: src/server/DownloadHandler.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        headers.set("Content-Type", "application/octet-stream");
+        headers.set("Content-Disposition", "attachment; filename="
+                + file.getName());
+        headers.set("Expires", "0");
+        headers.set("Cache-Control", "must-revalidate");
+        headers.set("Pragma", "public");
+        exchange.sendResponseHeaders(200, file.length());
+        OutputStream out = exchange.getResponseBody();
+
+        byte[] buf = new byte[8192];
```
Comment (tastyburger) [inline src/server/DownloadHandler.java]: You could probably use file.length() here too to make it dynamic.

File: src/server/Doorman.java (hunk 1 above)
Comment (23Skidoo) [inline src/server/Doorman.java]: I'll take a look. Should be in `ServerSettings`, though.

File: src/server/DownloadHandler.java (hunk 2 above)
Comment (23Skidoo) [inline src/server/DownloadHandler.java]: Keeping the buffer small allows to process large files in constant memory.

**Comment by 23Skidoo** [comment]
OK, it looks like the clients just pass in the absolute path on server, which is problematic from security purposes (`download.php?path=/etc/passwd`).

**Comment by 23Skidoo** [comment]
Added support for legacy uploads/downloads (that is, `upload?path=/some/absolute/path`). They will be removed eventually, so clients are encouraged to convert to new-style uploads/downloads (`POST $hostname/api/upload` and `GET $hostname/api/download/relative/path/to/file`).

**Comment by 23Skidoo** [comment]
Seems to work with the desktop client (after patching it a bit).

**Comment by 23Skidoo** [comment]
This is ready to be merged now, I think. Remaining issues got their own tickets: #167, #178, #179.

**Comment by skriez** [comment]
Ok, I'll look over it :)

**Comment by tastyburger** [comment]
LGTM.

**Pull request merged**